x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/notifications/AbstractAuditor.java

```diff
@@ -83,4 +168,21 @@ private XContentBuilder toXContentBuilder(ToXContent toXContent) {
             throw new RuntimeException(e);
         }
     }
+
+    private void writeBacklog() {
+        BulkRequest bulkRequest = new BulkRequest();
+        ToXContent doc = backlog.poll();
+        while (doc != null) {
+            bulkRequest.add(indexRequest(doc));
+            doc = backlog.poll();
+        }
+
+        client.bulk(bulkRequest, ActionListener.wrap(
+            bulkItemResponses -> {
+                backlog = null;
+                logger.trace("Successfully wrote audit message backlog after upgrading template");
```

Comment:
Is it possible that we empty the backlog, but then have to add something else to it?

I know this only happens on template changes, but it seems like setting it to null here is unnecessary and risks an NPE.

Reply:
Also, it would be good to log if there were failures. we could just write a warning, and then have the full failure stack be debug or something.

Reply:
I wanted to hit the NPE in tests and fail in a obvious way. If the logic is sound it should never fail. 

But yes it should be some sort of assertion so there is no chance of it in production code